### PR TITLE
Remove ip_addrs & mac_addrs fields if empty

### PR DESCRIPTION
**Pull request by apuntamb**
The hypervisor has empty arrays in `ip_addresses` and `mac_addresses` fields. This will fail validation on the inventory side and prevent the host record from appearing there. This has been handled already in the `foreman_rh_cloud` plugin under [foreman_rh_cloud/pull/249](https://github.com/theforeman/foreman_rh_cloud/pull/249). This PR removes both these fields when they have empty arrays as values.
The PR needs a rebase on top of [yupana/pull/323](https://github.com/quipucords/yupana/pull/323/files) once it's merged and I could then include these method calls in [_transform_single_host](https://github.com/quipucords/yupana/pull/323/files#diff-f049d03813e04872b2e244ca395a9aaaR269).

### CODE REVIEW

**Comment by patilsuraj767** [comment]
Tested pull-request changes on my local setup. Working as expected :+1:

File: yupana/processor/report_slice_processor.py [hunk 1]
```diff
@@ -192,6 +192,38 @@ def generate_upload_candidates(self):
                       for key in host.keys() if key not in ['cause', 'status_code']}
         return candidates
 
+    def _remove_ip_addresses(self, host: dict):
```
Comment (kgaikwad) [inline yupana/processor/report_slice_processor.py:195]: One can be confused by an existing method name as this method only removes `ip_addresses` key only when contains empty array or nothing.

```suggestion
    def _remove_empty_ip_addresses(self, host: dict):
```

File: yupana/processor/report_slice_processor.py [hunk 2]
```diff
@@ -192,6 +192,38 @@ def generate_upload_candidates(self):
...
+                del host['ip_addresses']
+                LOG.info(
+                    format_message(
+                        self.prefix,
+                        'Removed empty ip_addresses fact.',
+                        account_number=self.account_number,
+                        report_platform_id=self.report_platform_id
+                    )
+                )
+        return host
```
Comment (kgaikwad) [inline yupana/processor/report_slice_processor.py:209]:
```suggestion
        ip_addresses = host.get('ip_addresses')
        if ip_addresses is None or ip_addresses:
            return host

        del host['ip_addresses']
        LOG.info(
            format_message(
                self.prefix,
                'Removed empty ip_addresses fact.',
                account_number=self.account_number,
                report_platform_id=self.report_platform_id
            ))
        return host
```
What do you think?

File: yupana/processor/report_slice_processor.py [hunk 3]
```diff
@@ -192,6 +192,38 @@ def generate_upload_candidates(self):
...
+
+    def _remove_mac_addresses(self, host: dict):
```
Comment (kgaikwad) [inline yupana/processor/report_slice_processor.py:211]: ditto

**Review by kgaikwad — CHANGES_REQUESTED** [review]
Added few inline comments.

File: yupana/processor/report_slice_processor.py [hunk 4]
```diff
@@ -192,6 +192,38 @@ def generate_upload_candidates(self):
...
+        """Remove 'mac_addresses' field."""
+        if 'mac_addresses' in host:
```
Comment (ShimShtein) [inline yupana/processor/report_slice_processor.py:213]: Worth refactoring to short circuit, like the `ip_addresses` method.

**Review by ShimShtein — CHANGES_REQUESTED** [review]

File: yupana/processor/report_slice_processor.py (hunk 2 above)
Comment (apuntamb) [reply yupana/processor/report_slice_processor.py:209]: Thanks for reviewing @kgaikwad and @ShimShtein!
Why are we checking `if ip_addresses is None or ip_addresses` as one? Aren't they contradicting? Or am I missing something? I am guessing `ip_addresses is None` will give `false` even when the array will be empty [].
Sharing what I could try on a Python Interpreter:
``` python
>>> k = {'a': ['h']}
>>> b = k.get('a')
>>> b is None """output False as b = ['h']"""
>>> k = {'a': []}
>>> b = k.get('a')
>>> b is None  """output False as b = []"""
```
Thoughts?

Comment (kgaikwad) [reply yupana/processor/report_slice_processor.py:209]: @apuntamb,

As per my understanding, 
it should delete `ip_addresses` key when ip_addresses contains empty array i.e `host={'ip_addresses': []}` and 
for host values like `host={'ip_addresses': ['192.168.10.10']} or host={}`, it shouldn't execute delete key action.
When `host={'ip_addresses': []}`, `if condition` block won't run. 

If you think it is contradicting, you can ignore it. It was just a suggestion. :)

Comment (apuntamb) [reply yupana/processor/report_slice_processor.py:209]: @kgaikwad Your suggestion makes perfect sense. I think I didn't quite well catch up on the context of `None` here. Will apply these changes. :+1:

File: yupana/processor/report_slice_processor.py [hunk 5]
```diff
@@ -264,6 +296,8 @@ def _transform_single_host(self, host: dict):
         if 'system_profile' in host:
             host = self._transform_os_release(host)
             host = self._transform_os_kernel_version(host)
+            host = self._remove_ip_addresses(host)
+            host = self._remove_mac_addresses(host)
```
[the remark this answers is not in the document]
Comment (apuntamb) [reply yupana/processor/report_slice_processor.py]: done

File: yupana/processor/report_slice_processor.py (hunk 1 above)
Comment (apuntamb) [reply yupana/processor/report_slice_processor.py]: done.

File: yupana/processor/report_slice_processor.py (hunk 2 above)
Comment (apuntamb) [reply yupana/processor/report_slice_processor.py]: done.

File: yupana/processor/report_slice_processor.py (hunk 4 above)
Comment (apuntamb) [reply yupana/processor/report_slice_processor.py]: done.

**Review by kgaikwad — APPROVED** [review]
LGTM, thank you @apuntamb :+1: 
@ShimShtein, do you have any additional comments?

File: yupana/processor/tests_report_slice_processor.py [hunk 6]
```diff
@@ -690,3 +690,31 @@ def test_match_regex_and_find_version(self):
         os_version = self.processor._match_regex_and_find_version(
             host['system_profile']['os_release'])
         self.assertEqual(host_os_version, os_version)
+
+    def test_remove_empty_ip_addresses(self):
```
Comment (ShimShtein) [inline yupana/processor/tests_report_slice_processor.py:694]: For completeness sake, I would also add test to check `None` values:

``` py
        """Test remove host ip_addresses."""
        host = {}
        host = self.processor._remove_empty_ip_addresses(host)
        host = self.processor.test_remove_empty_mac_addresses(host)
        self.assertEqual(host, {})
```

Comment (apuntamb) [reply yupana/processor/tests_report_slice_processor.py:694]: done.

**Comment by apuntamb** [comment]
@ShimShtein @kgaikwad how about one final round of review now?

**Review by ShimShtein — APPROVED** [review]
:+1: LGTM

**Review by kgaikwad — APPROVED** [review]

**Comment by kgaikwad** [comment]
Thank you @apuntamb
